### PR TITLE
LibC+LibELF: Call destructors on `exit` and clean up crt0

### PR DESCRIPTION
**Pull request by spholz**
This PR makes `exit()` correctly call all destructors according to https://www.sco.com/developers/gabi/latest/ch5.dynamic.html#init_fini.

The `_init()` and `_fini()` in crt0.cpp and stdlib.cpp's `exit()` are unnecessary as we call them in the dynamic linker and we don't care about statically linked executables anymore.

### CODE REVIEW

**Comment by timschumi** [comment]
> we don't care about statically linked executables anymore.

Pretty sure they just have been low-priority so far.

We should definitely try and structure our initialization routines so that they at least try and work on more than "DynamicLoader with our LibC" (i.e. split everything into LibC, DynamicLoader and crt* as appropriate).

**Comment by supercomputer7** [comment]
> > we don't care about statically linked executables anymore.
> 
> Pretty sure they just have been low-priority so far.
> 
> We should definitely try and structure our initialization routines so that they at least try and work on more than "DynamicLoader with our LibC" (i.e. split everything into LibC, DynamicLoader and crt* as appropriate).

I feel like we should be able to support static linking, and in fact, we should probably easily allow users to build static binaries if they really want to, so I'd advocate to not remove crt* mechanisms to support such feature.

File: Userland/Libraries/LibELF/DynamicLinker.cpp [hunk 1]
```diff
@@ -594,6 +600,33 @@ static Result<void, DlErrorMessage> __dladdr(void const* addr, Dl_info* info)
...
+            (*fini_end)();
+        }
+    }
+
+    if (object->has_fini_section()) {
+        auto fini_function = object->fini_section_function();
+        (fini_function)();
+    }
+}
+
```
Comment (BertalanD) [inline Userland/Libraries/LibELF/DynamicLinker.cpp:629]: Shouldn't we also be calling the finalizers defined in other loaded objects as well, not just the main program (c.f. `DynamicLoader::call_object_init_functions`)?

**Comment by BertalanD** [comment]
> We should definitely try and structure our initialization routines so that they at least try and work on more than "DynamicLoader with our LibC" (i.e. split everything into LibC, DynamicLoader and crt* as appropriate).

I have tried locally to structure our crt objects in a way to allow that; and the fact is that DynamicLoader and our libc internals are intertwined to an extent to make it a huge task compared to the minor "benefit" it would bring. If someone wants to go through the trouble, feel free to do so, but I'd rather not block this PR on it.

> Sorry, I likely misunderstood @BertalanD. I thought he meant that we don't want to support them at all and that I should clean up crt0 accordingly.

It already does not work at all, and removing the broken "attempts" to support static linking from crt0 won't change anything.

File: Userland/Libraries/LibELF/DynamicLinker.cpp (hunk 1 above)
Comment (BertalanD) [reply Userland/Libraries/LibELF/DynamicLinker.cpp:629]: Would iterating in reverse topological order work? `OrderedHashMap` uses a doubly linked list internally, so it should be possible, we just need to implement a reverse iterator for it.

Comment (BertalanD) [reply Userland/Libraries/LibELF/DynamicLinker.cpp:629]: I could have a go tonight if you'd prefer.

Comment (BertalanD) [reply Userland/Libraries/LibELF/DynamicLinker.cpp:629]: #21219

File: Userland/Libraries/LibELF/DynamicLoader.cpp [hunk 2]
```diff
@@ -803,7 +803,7 @@ void DynamicLoader::call_object_init_functions()
     typedef void (*InitFunc)();
```
Comment (BertalanD) [inline Userland/Libraries/LibELF/DynamicLoader.cpp:803]: This `typedef` now seems unused.

**Review by BertalanD — COMMENTED** [review]
Tiny nit, otherwise LGTM!

Thanks for pushing this through and cleaning up this mess!

**Pull request merged**